execution_engine/src/system/auction.rs

```diff
@@ -528,93 +528,93 @@ pub trait Auction:
-            let recipient = seigniorage_recipients
-                .get(&public_key)
-                .ok_or(Error::ValidatorNotFound)?;
+        let recipient = seigniorage_recipients
+            .get(&proposer)
+            .ok_or(Error::ValidatorNotFound)?;
 
-            let total_stake = recipient.total_stake().ok_or(Error::ArithmeticOverflow)?;
-            if total_stake.is_zero() {
-                // TODO: error?
-                continue;
-            }
+        let total_stake = recipient.total_stake().ok_or(Error::ArithmeticOverflow)?;
+        /*if total_stake.is_zero() {
+            // TODO: error?
+            continue;
+        }*/
 
-            let total_reward: Ratio<U512> = {
-                let reward_rate = Ratio::new(U512::from(reward_factor), U512::from(BLOCK_REWARD));
-                reward_rate
-                    .checked_mul(&Ratio::from(base_round_reward))
-                    .ok_or(Error::ArithmeticOverflow)?
-            };
+        let total_reward: Ratio<U512> = Ratio::from(base_round_reward);
 
-            let delegator_total_stake: U512 = recipient
-                .delegator_total_stake()
-                .ok_or(Error::ArithmeticOverflow)?;
+        let delegator_total_stake: U512 = recipient
+            .delegator_total_stake()
+            .ok_or(Error::ArithmeticOverflow)?;
 
-            let delegators_part: Ratio<U512> = {
-                let commission_rate = Ratio::new(
-                    U512::from(*recipient.delegation_rate()),
-                    U512::from(DELEGATION_RATE_DENOMINATOR),
-                );
-                let reward_multiplier: Ratio<U512> = Ratio::new(delegator_total_stake, total_stake);
-                let delegator_reward: Ratio<U512> = total_reward
-                    .checked_mul(&reward_multiplier)
-                    .ok_or(Error::ArithmeticOverflow)?;
-                let commission: Ratio<U512> = delegator_reward
-                    .checked_mul(&commission_rate)
-                    .ok_or(Error::ArithmeticOverflow)?;
-                delegator_reward
-                    .checked_sub(&commission)
-                    .ok_or(Error::ArithmeticOverflow)?
-            };
+        let delegators_part: Ratio<U512> = {
+            let commission_rate = Ratio::new(
+                U512::from(*recipient.delegation_rate()),
+                U512::from(DELEGATION_RATE_DENOMINATOR),
+            );
+            let reward_multiplier: Ratio<U512> = Ratio::new(delegator_total_stake, total_stake);
+            let delegator_reward: Ratio<U512> = total_reward
+                .checked_mul(&reward_multiplier)
+                .ok_or(Error::ArithmeticOverflow)?;
+            let commission: Ratio<U512> = delegator_reward
+                .checked_mul(&commission_rate)
+                .ok_or(Error::ArithmeticOverflow)?;
+            delegator_reward
+                .checked_sub(&commission)
+                .ok_or(Error::ArithmeticOverflow)?
+        };
```

Comment:
(This is not a new issue, of course, but:)
We should make sure that and document why these errors are unreachable.

I think both `delegator_reward` and `commission` have at most 64 * 2 = 128 resp. 64 * 3 = 192 bits in the denominator and numerator, so after the subtraction the numerator has at most 320, and will then be reduced back to at most 192. (If I understand correctly, the crate automatically reduces fractions?) In `delegator_rewars` another 64 bits are added (⇒ ≤ 256). `reinvest_delegator_rewards` seems to immediately round them to an integer, so that should be fine — would feel a bit safer to round them here and not use `Ratio` in `reinvest_delegator_rewards`.

And I guess the subtraction here can't fail because we don't allow `commission`s greater than 100%? (Do we check that?) It might be safer to just `saturating_sub` here?

Reply:
we can and should reevaluate this, but Alex doesn't need to be blocked on it for this PR